lib/src/shortest_path.dart

```diff
@@ -10,56 +10,94 @@ import 'dart:collection';
 /// If [start] `==` [target], an empty [List] is returned and [edges] is never
 /// called.
 ///
-/// [V] is the type of values in the graph nodes. [K] must be a type suitable
-/// for using as a Map or Set key, and [key] must provide a consistent key for
-/// every node.
-///
 /// [start], [target] and all values returned by [edges] must not be `null`.
 /// If asserts are enabled, an [AssertionError] is raised if these conditions
 /// are not met. If asserts are not enabled, violations result in undefined
 /// behavior.
-List<V> shortestPath<K, V>(
-        V start, V target, K Function(V) key, Iterable<V> Function(V) edges) =>
-    _shortestPaths<K, V>(start, key, edges, key(target))[key(target)];
+///
+/// If [equals] is provided, it is used to compare nodes in the graph. If
+/// [equals] is omitted, the node's own [Object.==] is used instead.
+///
+/// Similarly, if [hashCode] is provided, it is used to produce a hash value
+/// for nodes to efficiently calculate the return value. If it is omitted, the
+/// key's own [Object.hashCode] is used.
+///
+/// If you supply one of [equals] or [hashCode], you should generally also to
+/// supply the other.
+List<T> shortestPath<T>(
+  T start,
+  T target,
+  Iterable<T> Function(T) edges, {
+  bool equals(T key1, T key2),
+  int hashCode(T key),
+}) =>
+    _shortestPaths<T>(
+      start,
+      edges,
+      target: target,
+      equals: equals,
+      hashCode: hashCode,
+    )[target];
 
 /// Returns a [Map] of the shortest paths from [start] to all of the nodes in
 /// the directed graph defined by [edges].
 ///
 /// All return values will contain the key [start] with an empty [List] value.
 ///
-/// [V] is the type of values in the graph nodes. [K] must be a type suitable
-/// for using as a Map or Set key, and [key] must provide a consistent key for
-/// every node.
-///
 /// [start] and all values returned by [edges] must not be `null`.
 /// If asserts are enabled, an [AssertionError] is raised if these conditions
 /// are not met. If asserts are not enabled, violations result in undefined
 /// behavior.
-Map<K, List<V>> shortestPaths<K, V>(
-        V start, K Function(V) key, Iterable<V> Function(V) edges) =>
-    _shortestPaths<K, V>(start, key, edges);
+///
+/// If [equals] is provided, it is used to compare nodes in the graph. If
+/// [equals] is omitted, the node's own [Object.==] is used instead.
+///
+/// Similarly, if [hashCode] is provided, it is used to produce a hash value
+/// for nodes to efficiently calculate the return value. If it is omitted, the
+/// key's own [Object.hashCode] is used.
+///
+/// If you supply one of [equals] or [hashCode], you should generally also to
+/// supply the other.
+Map<T, List<T>> shortestPaths<T>(
+  T start,
+  Iterable<T> Function(T) edges, {
+  bool equals(T key1, T key2),
```

Comment:
You can consider using a single `Equality` from `package:collection` instead of passing around separate equals and hash-code functions.

Reply:
It's not a requirement but we've been attempting to keep this package to no dependencies. I also think I prefer consistency with the core libraries.

Reply:
I thought about using `Equality`,too – but avoiding other deps sounded nicer.